business 로직 재정의

k-ye0415 opened the pull request:
## 숙제

High-level modules should not depend on low-level modules. Both should depend on abstractions.
위 규칙의 위반입니다.
누가 보더라도 high-level module(repository)이 low-level module(api service)에 의존하고 있죠?
이를 뒤집어서, repository의 인터페이스를 low-level에서 구현하도록 만들어줘야 합니다.
NG부분: "PlantRepository"는 상위 레이어의 domain logic 입니다. 이 녀석의 interface는 "data" layer에 있어선 안된다는 거죠. 지금은 "data" 레이어이 있는 듯 하네요.
OK부분: "PlantRepositoryImpl"는 "detail"입니다. "detail"은 "data" layer에서 "UnsplashApiService"의 abstraction만 가지고 구현해야 합니다.

------
- UI layer
- Domain layer
  - UseCase(GetPlantListUseCase)
  - ~~Interface~~
    - Repository Interface
    - ~~UnsplashService, Response~~ API 통신하는 부분은 다 Data Layer 이다.
    - ~~WikipediaService, Response~~ API 통신하는 부분은 다 Data Layer 이다.
  - Entity
- Data layer
  - Repository Implementation
  - Unsplash
  - Wikipedia

Commit:
business 로직 재정의

- UI layer
- Domain layer
  - UseCase(GetPlantListUseCase)
  - Interface
    - Repository
    - UnsplashService, Response
    - WikipediaService, Response
- Data layer
  - Repository Implementation
  - API 호출 부

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/domain/unsplash/UnsplashService.kt:
```diff
@@ -0,0 +1,23 @@
+    companion object {
+        val unsplashApi: UnsplashService = Retrofit.Builder()
+            .baseUrl("https://api.unsplash.com/")
+            .addConverterFactory(GsonConverterFactory.create())
```
gson의 성능은 얼마나 gson 내부의 모델 캐시를 재활용하는가에 달려있습니다. 즉, gson instance를 얼마만큼 재활용하는가가 중요한 요소이지요. 지금은 service factory는 gson을 개별적으로 생성하고 있어서 비효율적입니다. 어떻게 하면 좋을까요?

k-ye0415 replied:
`GsonConverterFactory.create()`를 사용하면 내부적으로 새로운 Gson 객체를 계속 생성하기에 따로 인스턴스를 정의 후 재사용할 수 있도록 개선했습니다.

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/domain/repository/PlantRepository.kt:
```diff
@@ -0,0 +1,8 @@
+
+interface PlantRepository {
+    suspend fun getPlantList(plant: Plant): Plant
+    suspend fun loadPlantList(): List<Plant>
```
둘 다 plant list를 얻어오는 것 같은데 하나는 argument를 요구하고 또 다른 하나는 요구하지 않는 이유가 이해가 안되요.

예진님은 알고 계시지만 인터페이스에는 녹여져있지 않은 맥락이 있을지도 모르겠어요.

jerry-f-lab replied:
여기도 마찬가지이지만 아마도 getPlantList가 simple getter가 아니기 때문에 argument를 요구하는 것이겠죠?

k-ye0415 replied:
또.. 저만.. 아는... 비밀스럽고 배려심없는 코드를 작성해버렷네요;;... 인성이 문제가 있나봅니다 😰

jerry-f-lab replied:
무슨 그런 심한 말씀까지 😢
전혀 그렇게 생각하지 않았어요.

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/data/wikipedia/WikipediaDataSource.kt:
```diff
@@ -0,0 +1,15 @@
+import com.jin.sunflower.core.domain.wikipedia.WikipediaService
+
+class WikipediaDataSource(private val apiService: WikipediaService) {
+    suspend fun getPlantDescription(query: String): String {
```
파라메터명과 함수명이 아주 잘 어울리지는 않아보여요. query라는 명사는, search, lookup, find, retrieve등의 동사와 잘 어울리고, get과는 잘 아울리지 않아요.

jerry-f-lab replied:
get은 대단히 뜻이 많은 다의어이기 때문에 의미를 모호하게 만드는 경향이 강합니다. 지극히 단순한 getter를 제외하고는 get이란 동사를 함수명에 사용하는 것을 피하세요. 지극한 단순한 getter(simple getter)가 무엇인지는 공부해보세요.

k-ye0415 replied:
`get`이란 너무 광범위한 뜻을 지녔기 때문에 각 상황에 맞게끔 작성하는게 좋다.
특히 프로퍼티의 `get`과 혼동될 수 있기에 더더욱 조심해야한다.
네트워크 : fecth, retrieve ..
데이터 변환 : convert, format ...
조건 검색 : search, find ...
엑세스 : access, read
내부 가져오기 : load

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/data/unsplash/UnsplashDataSource.kt:
```diff
@@ -0,0 +1,21 @@
+
+class UnsplashDataSource(private val apiService: UnsplashService) {
+
+    suspend fun getImageUrl(query: String): String {
```
이쯤 되면 예진님이 얼마나 get을 좋아하시는지 아시겠죠?😹
동의어사전(thesaurus)을 보면서 어떤 동사를 사용해야 가장 이 함수를 잘 설명할 수 있을지 생각해보세요. 다른 getXxx함수들도 마찬가지 입니다.

Commit:
불필요한 객체 생성 방지하기 위해 하나의 Gson 객체를 사용할 수 있도록 수정.

Commit:
메소드 명 개선

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/domain/RetrofitClient.kt:
```diff
@@ -0,0 +1,22 @@
+object RetrofitClient {
+    private val gson: Gson = GsonBuilder().create()
+
+    fun <T> createService(baseUrl: String, service: Class<T>): T {
```
추가: reified inline 함수를 이용해서 service parameter를 없애봅시다.

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/domain/RetrofitClient.kt:
```diff
@@ -0,0 +1,22 @@
+import retrofit2.converter.gson.GsonConverterFactory
+
+object RetrofitClient {
+    private val gson: Gson = GsonBuilder().create()
```
nice.

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/domain/RetrofitClient.kt:
```diff
@@ -0,0 +1,22 @@
+package com.jin.sunflower.core.domain
```
이 클래스는 왜 domain에 있는 걸까요?
비즈니스와 관련되어 있나요?

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/domain/unsplash/UnsplashResponse.kt:
```diff
@@ -1,4 +1,4 @@
-package com.jin.sunflower.core.model
+package com.jin.sunflower.core.domain.unsplash
```
이것도 그렇네요.
domain layer라 함은 비즈니스의 도메인을 표현하는 레이어입니다.
비즈니스의 도메인이라 함은 "결제", "주문", "배송", "친구 관계" 와 같은 어플리케이션의 근간을 이루는 기능적 추상화입니다.

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/domain/wikipedia/WikipediaService.kt:
```diff
@@ -0,0 +1,19 @@
+package com.jin.sunflower.core.domain.wikipedia
```
이것도 그렇네요.
이 어플리케이션의 비즈니스 도메인이 위키피디아인가요?

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/domain/repository/PlantRepository.kt:
```diff
@@ -0,0 +1,8 @@
+
+interface PlantRepository {
+    suspend fun loadPlantList(): List<Plant>
+    suspend fun fetchPlantData(plant: Plant): Plant
```
함수 시그니쳐가 변경되고 나니까 또다시 이상한점이 눈에 띄는데요. Plant를 주고 fetch를 요청하면 Plant가 되돌아오네요. 그리고 repository는 비즈니스 로직이 있어야 하는데 아직도 fetching data를 하고 있군요.

Commit:
`inline fun <reified T>` 함수 재정의 및 디렉토리 변경

Commit:
메소드 명 재정의 및 usecase 수정

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/domain/repository/PlantRepository.kt:
```diff
@@ -0,0 +1,8 @@
+
+interface PlantRepository {
+    suspend fun loadPlantList(): List<Plant>
+    suspend fun searchPlantByName(plant: Plant): Plant
```
searchByPlantName인데도 플랜트가 들어가서 플랜트가 나오네요. 
plantName이 들어가서 Plant?가 나와야 할 것 같은데 말이죠.

k-ye0415 replied:
name 으로 검색하는데, plant 객체로 넘기고, 또 같은 객체를 반환하는 것이 잘못됬음을 충분히 이해했습니다. (🫣부끄..)

description, imageUrl 외 모두 다 하드코딩 되고 있는 부분인데 미리 list로 정의하는 것을 고집을 부렸네요..;;
id 같은 경우에는 현재 사용하지도 않고, 하드코딩하기에 적절하지않은 것 같아서 (plantNameList index 로 할 수 있었겠지만..)
id 는 삭제하고 나머지 부분만 기존처럼 하드코딩 했습니다.

jerry-company replied:
인터페이스는 구현 편의성에 맞추어서 선언하는 것이 아닙니다. 오히려 함수를 호출하는 쪽의 구현 편의성에 맞추어서 선언하는 것이 올바른 방향이라고 할까요?

제 생각에는 repository impl안에 있던 리스트를 지울 필요는 없었을 것 같아요.
'list.map { searchByPlantName(it.name) }.filterNotNull()' 비슷하게 하면 loadPlantList()를 구현할 수 있었을 텐데 말이죠.
구현은 인터페이스에 맞춰서 그냥 하는 것 입니다.

jerry-f-lab replied:
인터페이스는 구현 편의성에 맞추어서 선언하는 것이 아닙니다. 오히려 함수를 호출하는 쪽의 구현 편의성에 맞추어서 선언하는 것이 올바른 방향이라고 할까요?

제 생각에는 repository impl안에 있던 리스트를 지울 필요는 없었을 것 같아요.
'list.map { searchByPlantName(it.name) }.filterNotNull()' 비슷하게 하면 loadPlantList()를 구현할 수 있었을 텐데 말이죠.
마찬가지로 'list.findOrNull { it.name == plantName }...' 같이 하면 searchByPlantName도 구현할 수 있었겠죠?
구현은 인터페이스에 맞춰서 그냥 하는 것 입니다.

Commit:
데이터클래스, mockData 수정 및 Repository 수정

- description, imageUrl 외 모두 다 하드코딩 되고 있는 부분인데 미리 list로 정의하는 것을 고집부림(?).
- name 만 list 로 정의 후 처리할 수 있도록 수정.

jerry-f-lab reviewed (approved):
LGTM
